Update `compiler-builtins` and enable f128 tests on all non-buggy platforms

tgross35 opened the pull request:
Update compiler_builtins to 0.1.138 and pin it. This updates to a new version of builtins that includes [1], which was
the last blocker to us enabling `f128` tests on all platforms.

With that, we now provide symbols necessary to work with `f128` everywhere. This means that we are no longer restricted to systems that provide `f128` symbols themselves, and can enable tests by default.

There are still a handful of platforms that need to remain disabled because of bugs and some that had to get updated.

Math support is still off by default since those symbols are not yet available.

[1]: https://github.com/rust-lang/compiler-builtins/pull/624

try-job: test-various
try-job: i686-gnu-nopt

tgross35 committed:
Update `compiler_builtins` to 0.1.138 and pin it

This updates to a new version of builtins that includes [1], which was
the last blocker to us enabling `f128` tests on all platforms 🎉.

With this update, also change to pinning the version with `=` rather
than using the default carat versioning. This is meant to ensure that
`compiler-builtins` does not get updated as part of the weekly
`Cargo.lock` update, since updates to this crate need to be intentional:
changes to rust-lang/rust and rust-lang/compiler-builtins sometimes need
to be kept in lockstep, unlike most dependencies, and sometimes these
updates can be problematic.

[1]: https://github.com/rust-lang/compiler-builtins/pull/624

tgross35 committed:
Enable `f128` tests on all non-buggy platforms 🎉

With the `compiler-builtins` update to 0.1.137 [1], we now provide
symbols necessary to work with `f128` everywhere. This means that we are
no longer restricted to 64-bit linux, and can enable tests by default.

There are still a handful of platforms that need to remain disabled
because of bugs. This patch additionally disables the following:

1. MIPS [2]
2. 32-bit x86 [3]

Math support is still off by default since those symbols are not yet
available.

[1]: https://github.com/rust-lang/rust/pull/132433
[2]: https://github.com/llvm/llvm-project/issues/96432
[3]: https://github.com/llvm/llvm-project/issues/77401

tgross35 commented:
This includes https://github.com/rust-lang/rust/pull/132433 which includes https://github.com/rust-lang/rust/pull/132206, so the chain is a bit deep (I'm keeping updates separate from rust-lang/rust changes so we aren't blocked from updating further if something goes wrong here). But that doesn't mean we can't start some tests.

@bors try

tgross35 commented:
Those are quite a few try jobs but it still doesn't cover everything that will be newly enabled here.

@bors rollup=never

bors commented:
:hourglass: Trying commit 15075bf74a7a7112db3422634f4f3a0974dcc02f with merge ec0220aadb3398fa8aa480a8d4a0f8e3e6d48e20...
<!-- homu: {"type":"TryBuildStarted","head_sha":"15075bf74a7a7112db3422634f4f3a0974dcc02f","merge_sha":"ec0220aadb3398fa8aa480a8d4a0f8e3e6d48e20"} -->

tgross35 commented:
Also cc @beetrees

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11622131232/job/32370546755)
Build commit: ec0220aadb3398fa8aa480a8d4a0f8e3e6d48e20 (`ec0220aadb3398fa8aa480a8d4a0f8e3e6d48e20`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11622131232/job/32370546755"},"merge_sha":"ec0220aadb3398fa8aa480a8d4a0f8e3e6d48e20"} -->

workingjubilee commented:
This has a merge conflict, it seems.

workingjubilee commented:
I have rebased it.

workingjubilee commented:
I do not believe we run tests on mips64

workingjubilee commented:
I decided to change up the suite of tests slightly and rerun this. In particular, I noticed we didn't run the infamous `test-various`.

@bors try

bors commented:
:hourglass: Trying commit 95998eebd479975acdb609e3445ca42199f9237e with merge ad17ccbd86163949fa3f4f8c1a9ff6527cc08b24...
<!-- homu: {"type":"TryBuildStarted","head_sha":"95998eebd479975acdb609e3445ca42199f9237e","merge_sha":"ad17ccbd86163949fa3f4f8c1a9ff6527cc08b24"} -->

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/11655287261/job/32449788895) [(plain)](https://github.com/rust-lang-ci/rust/commit/ad17ccbd86163949fa3f4f8c1a9ff6527cc08b24/checks/32449788895/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test-various
dist-mips-linux
dist-mips64el-linux
##[endgroup]
INFO:root:Job type: TryRunType(custom_jobs=['armhf-gnu', 'i686-gnu', 'i686-gnu-nopt', 'test-various', 'dist-mips-linux', 'dist-mips64el-linux'])
  File "/home/runner/work/rust/rust/src/ci/github-actions/calculate-job-matrix.py", line 195, in <module>
  File "/home/runner/work/rust/rust/src/ci/github-actions/calculate-job-matrix.py", line 195, in <module>
    jobs = calculate_jobs(run_type, data)
  File "/home/runner/work/rust/rust/src/ci/github-actions/calculate-job-matrix.py", line 138, in calculate_jobs
    raise Exception(f"Custom job(s) `{unknown_jobs}` not found in auto jobs")
Exception: Custom job(s) `['dist-mips-linux', 'dist-mips64el-linux']` not found in auto jobs
##[error]Process completed with exit code 1.
```

</details>

tgross35 commented:
I updated this to skip x86 and mips as mentioned. Indeed we don't test this (and try job only allow the jobs listed [here](https://github.com/rust-lang-ci/rust/blob/b8c8287a229cd79604aa84c25e1235fc78cd5f2e/src/ci/github-actions/jobs.yml#L118), though I wish it was possible to do try jobs for tier 3 targets) but I know there are people running the test suite locally, e.g. https://github.com/rust-lang/rust/pull/127235.

Updated try jobs to those that have been mentioned but not yet tested.

@bors try

bors commented:
:hourglass: Trying commit 898ecf0a04e5985a00925419766b4b015b1afc39 with merge 2ec481c395b83f7d7eb0b3521f6ab25f23a6360d...
<!-- homu: {"type":"TryBuildStarted","head_sha":"898ecf0a04e5985a00925419766b4b015b1afc39","merge_sha":"2ec481c395b83f7d7eb0b3521f6ab25f23a6360d"} -->

tgross35 commented on library/std/build.rs:
```diff
@@ -122,19 +122,24 @@ fn main() {
+        ("mips" | "mips64" | "mips32r6" | "mips64r6", _) => false,
         // Selection bug <https://github.com/llvm/llvm-project/issues/95471>
         ("nvptx64", _) => false,
+        // ABI and precision bugs <https://github.com/rust-lang/rust/issues/125109> et al. (full
```
Thanks, updated both of these

beetrees reviewed (approved):
LGTM. Thanks

workingjubilee commented:
> All MIPS targets are tier 3, so no builds or tests of any kind are run by rust-lang/rust. The benefit of adding it to the match statement is just for those running tests locally/Linux distros with a MIPS64 port etc. 32-bit sparc is also tier 3 only and is already on the list.

mmh. my usual inclination is that we should leave known-failing-on-tier-3 tests on, so that people who are doing _just that_ are aware that _the platform's support is incomplete_... and if they run the tests and then are bothered by it, we're usually amenable to exclusions.

as this is a nightly feature though I'm not strongly motivated either way.

workingjubilee commented:
let's get this show on the road:
@bors r+

bors commented:
:pushpin: Commit c0fc25cc204ba0a9ff4e78a2bcc1ebd0d05a995f has been approved by `workingjubilee`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=workingjubilee c0fc25cc204ba0a9ff4e78a2bcc1ebd0d05a995f -->
<!-- homu: {"type":"Approved","sha":"c0fc25cc204ba0a9ff4e78a2bcc1ebd0d05a995f","approver":"workingjubilee","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c0fc25cc204ba0a9ff4e78a2bcc1ebd0d05a995f with merge 706eec8ce1127c9970c8586eeb47a2966e4ddff4...
<!-- homu: {"type":"BuildStarted","head_sha":"c0fc25cc204ba0a9ff4e78a2bcc1ebd0d05a995f","merge_sha":"706eec8ce1127c9970c8586eeb47a2966e4ddff4"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/11657858467/job/32459665727)
Approved by: workingjubilee
Pushing 706eec8ce1127c9970c8586eeb47a2966e4ddff4 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"workingjubilee","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/11657858467/job/32459665727"},"merge_sha":"706eec8ce1127c9970c8586eeb47a2966e4ddff4"} -->

rust-timer commented:
Finished benchmarking commit (706eec8ce1127c9970c8586eeb47a2966e4ddff4): [comparison URL](https://perf.rust-lang.org/compare.html?start=43c78051ea96f89ecf33533dfed6db6f1eeba150&end=706eec8ce1127c9970c8586eeb47a2966e4ddff4&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=43c78051ea96f89ecf33533dfed6db6f1eeba150&end=706eec8ce1127c9970c8586eeb47a2966e4ddff4&stat=instructions:u)
This is the most reliable metric that we have; it was used to determine the overall result at the top of this comment. However, even this metric can sometimes exhibit noise.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.5%  | [1.5%, 1.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.4% | [-1.4%, -1.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=43c78051ea96f89ecf33533dfed6db6f1eeba150&end=706eec8ce1127c9970c8586eeb47a2966e4ddff4&stat=max-rss)
<details>
<summary>Results (primary -1.2%, secondary 1.4%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 2.8%  | [1.2%, 4.8%]   | 3     |
| Improvements ✅ <br /> (primary)   | -1.2% | [-1.2%, -1.2%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.0% | [-3.0%, -3.0%] | 1     |
| All ❌✅ (primary)                 | -1.2% | [-1.2%, -1.2%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=43c78051ea96f89ecf33533dfed6db6f1eeba150&end=706eec8ce1127c9970c8586eeb47a2966e4ddff4&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=43c78051ea96f89ecf33533dfed6db6f1eeba150&end=706eec8ce1127c9970c8586eeb47a2966e4ddff4&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 780.005s -> 778.282s (-0.22%)
**Artifact size**: 335.27 MiB -> 335.31 MiB (0.01%)

jethrogb commented:
Is it correct that compiler-builtins doesn't define `fmodl` (for f128)?

jethrogb commented:
I see in your first message you mention:

> Math support is still off by default since those symbols are not yet available.

However, the compiler does appear to be generating references to `fmodl` with this PR merged.
